### ISSUE TITLE
Error on Restarting Celery

### ISSUE THREAD

**Issue by AlecRosenbaum** [celery/kombu#970]
After upgrading from `kombu==4.2.1` to `kombu==4.2.2`, I'm now seeing this error while doing warm celery shutdowns. Unacked messages don't end up being restored and tasks are lost.

Downgrading to `kombu==4.2.1` fixes the issue for me.

```
worker: Warm shutdown (MainProcess)
[2018-12-19 13:16:30,578: WARNING/MainProcess] Restoring 1 unacknowledged message(s)
[2018-12-19 13:16:30,578: WARNING/MainProcess] Traceback (most recent call last):
[2018-12-19 13:16:30,578: WARNING/MainProcess] File "/usr/local/lib/python2.7/multiprocessing/util.py", line 274, in _run_finalizers
[2018-12-19 13:16:30,578: WARNING/MainProcess] finalizer()
[2018-12-19 13:16:30,579: WARNING/MainProcess] File "/usr/local/lib/python2.7/multiprocessing/util.py", line 207, in __call__
[2018-12-19 13:16:30,579: WARNING/MainProcess] res = self._callback(*self._args, **self._kwargs)
[2018-12-19 13:16:30,579: WARNING/MainProcess] File "/usr/local/lib/python2.7/site-packages/kombu/transport/virtual/base.py", line 287, in restore_unacked_once
[2018-12-19 13:16:30,579: WARNING/MainProcess] unrestored = self.restore_unacked()
[2018-12-19 13:16:30,579: WARNING/MainProcess] File "/usr/local/lib/python2.7/site-packages/kombu/transport/redis.py", line 164, in restore_unacked
[2018-12-19 13:16:30,580: WARNING/MainProcess] with self.channel.conn_or_acquire(client) as client:
[2018-12-19 13:16:30,580: WARNING/MainProcess] File "/usr/local/lib/python2.7/contextlib.py", line 17, in __enter__
[2018-12-19 13:16:30,580: WARNING/MainProcess] return self.gen.next()
[2018-12-19 13:16:30,580: WARNING/MainProcess] File "/usr/local/lib/python2.7/site-packages/kombu/transport/redis.py", line 978, in conn_or_acquire
[2018-12-19 13:16:30,581: WARNING/MainProcess] yield self._create_client()
[2018-12-19 13:16:30,582: WARNING/MainProcess] File "/usr/local/lib/python2.7/site-packages/kombu/transport/redis.py", line 959, in _create_client
[2018-12-19 13:16:30,582: WARNING/MainProcess] return self.Client(connection_pool=self.pool)
[2018-12-19 13:16:30,582: WARNING/MainProcess] File "/usr/local/lib/python2.7/site-packages/kombu/transport/redis.py", line 983, in pool
[2018-12-19 13:16:30,583: WARNING/MainProcess] self._pool = self._get_pool()
[2018-12-19 13:16:30,583: WARNING/MainProcess] File "/usr/local/lib/python2.7/site-packages/kombu/transport/redis.py", line 962, in _get_pool
[2018-12-19 13:16:30,583: WARNING/MainProcess] params = self._connparams(asynchronous=asynchronous)
[2018-12-19 13:16:30,583: WARNING/MainProcess] File "/usr/local/lib/python2.7/site-packages/kombu/transport/redis.py", line 888, in _connparams
[2018-12-19 13:16:30,584: WARNING/MainProcess] 'host': conninfo.hostname or '127.0.0.1',
[2018-12-19 13:16:30,584: WARNING/MainProcess] AttributeError: 'NoneType' object has no attribute 'hostname'
```

Package versions:

```
kombu==4.2.2
celery==4.2.1
redis==2.10.6
billiard==3.5.0.5
```

**Comment by AlecRosenbaum** [comment]
Not sure why this auto-resolved....

**Comment by vdmit11** [comment]
I've got the same problem:

```
celery_worker    | worker: Warm shutdown (MainProcess)
celery_worker    | Restoring 27 unacknowledged message(s)
celery_worker    | Traceback (most recent call last):
celery_worker    |   File "/usr/lib/python2.7/multiprocessing/util.py", line 274, in _run_finalizers
celery_worker    |     finalizer()
celery_worker    |   File "/usr/lib/python2.7/multiprocessing/util.py", line 207, in __call__
celery_worker    |     res = self._callback(*self._args, **self._kwargs)
celery_worker    |   File "/usr/local/lib/python2.7/dist-packages/kombu/transport/virtual/base.py", line 287, in restore_unacked_once
celery_worker    |     unrestored = self.restore_unacked()
celery_worker    |   File "/usr/local/lib/python2.7/dist-packages/kombu/transport/redis.py", line 164, in restore_unacked
celery_worker    |     with self.channel.conn_or_acquire(client) as client:
celery_worker    |   File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
celery_worker    |     return self.gen.next()
celery_worker    |   File "/usr/local/lib/python2.7/dist-packages/kombu/transport/redis.py", line 978, in conn_or_acquire
celery_worker    |     yield self._create_client()
celery_worker    |   File "/usr/local/lib/python2.7/dist-packages/kombu/transport/redis.py", line 959, in _create_client
celery_worker    |     return self.Client(connection_pool=self.pool)
celery_worker    |   File "/usr/local/lib/python2.7/dist-packages/kombu/transport/redis.py", line 983, in pool
celery_worker    |     self._pool = self._get_pool()
celery_worker    |   File "/usr/local/lib/python2.7/dist-packages/kombu/transport/redis.py", line 962, in _get_pool
celery_worker    |     params = self._connparams(asynchronous=asynchronous)
celery_worker    |   File "/usr/local/lib/python2.7/dist-packages/kombu/transport/redis.py", line 888, in _connparams
celery_worker    |     'host': conninfo.hostname or '127.0.0.1',
celery_worker    | AttributeError: 'NoneType' object has no attribute 'hostname'
celery_worker exited with code 0
```

Happens when trying to shutdown a Celery worker that has scheduled (but not yet executed) tasks.

And I can confirm that this bug was introduced in v4.2.2, and downgrading to v4.2.1 solves the problem.

**Comment by ntravis** [comment]
see #966 (#963 is basically the same issue as yours that 4.2.2 was master, not just 4.2 branch)

**Comment by vdmit11** [comment]
After some play with `git bisect`, I found the first commit where it happens: 9b0694f4ed52f325429ec3e1b276ed55bafc8b59

After a bit of investigation, I found one thing about this change: it changes the order of operations inside the `disconnect()` method.

That is, previously, the code looked like this:

```python
               def disconnect(self):
                    super(Connection, self).disconnect()
                    channel._on_connection_disconnect(self)
```

And then the author replaced `super()` with the copy-pasted and slightly modified version of the code from the base `redis.Connection` class:

```python
    def disconnect(self):
        "Disconnects from the Redis server"
        self._parser.on_disconnect()
        if self._sock is None:
            return
        try:
            self._sock.shutdown(socket.SHUT_RDWR)
            self._sock.close()
        except socket.error:
            pass
        self._sock = None
```

And the combined version looks like this:

```python
                def disconnect(self):
                    # NOTE: see celery issue #3898
                    # redis-py Connection shutdown()s the socket
                    # which causes all copies of file descriptor
                    # to become unusable, however close() only
                    # affect process-local copies of fds.
                    # So we just override Connection's disconnect method.
                    self._parser.on_disconnect()
                    channel._on_connection_disconnect(self)
                    if self._sock is None:
                        return
                    try:
                        # self._sock.shutdown(socket.SHUT_RDWR)
                        self._sock.close()
                    except socket.error:
                        pass
                    self._sock = None
```

And the intention here was to just comment-out the `sock.shutdown()` call.

But, there is one extra change here (not sure if that was intentional): the `channel._on_connection_disconnect(self)` line previously was below `super()` call, and now it is somewhere in the middle of the code that came from `super()`.

And if you move this `_on_connection_disconnect()` call to the bottom of the method, it actually fixes the bug.

But, I am not sure if this is a good fix, because I looked inside this `_on_connection_disconnect()`, and found, that it should un-register the socket in epoll (event poll, linux-specific), but actually that doesn't happen, because by the moment the method is called, the socket is already closed. 

Long story short, when placed in the middle, the `_on_connection_disconnect()` does its job, and when placed at the end of the method body, the call has no effect.

So previously, this call also had no effect. And to me, it looks like something was always broken here, and the commit fixed this broken thing. And because it now works well and cleans up the connection properly, it broke some other code (the `_connparams()` method?) that relies on some connection object that is now deleted.

But, I am not very confident here, because it is the very first time I'm even looking to the source code of the Kombu project. So I would like to leave further investigation to people who have more experience here.

**Comment by vdmit11** [comment]
Ok, maybe this is not quite for v4.2.2 (see #966), but the problem still exists in the master branch, and you will eventually have to fix it.

**Comment by psgibbs** [comment]
@vdmit11 I'm curious if you've seen https://github.com/celery/kombu/commit/eb6e4c8d512c6451d2430744322014439e8a0df8

It seems like this reverts the original change, and fixes that issue with a dependency update, meanwhile moving `channel._on_connection_disconnect(self)` back to where it was originally.

Are people still experiencing this?

**Comment by vdmit11** [comment]
> @vdmit11 I'm curious if you've seen [eb6e4c8](https://github.com/celery/kombu/commit/eb6e4c8d512c6451d2430744322014439e8a0df8)
> 
> It seems like this reverts the original change, and fixes that issue with a dependency update, meanwhile moving `channel._on_connection_disconnect(self)` back to where it was originally.
> 
> Are people still experiencing this, because it seems to be resolved

I didn't try it yet, but I think that should fix this specific bug with shutdown.

But as I mentioned earlier, to me that looks as a bad fix, because in this case the `_on_connection_disconnect()` call just has no effect.
So the bug is temporary masked by this fact that the function is called at a wrong time (when the socket is already closed), and it just does nothing and returns immediately, not letting the buggy code to be executed.
So in future, this bug may appear again.
